src/database-postgresql.cpp

```diff
@@ -0,0 +1,238 @@
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "config.h"
+
+#if USE_POSTGRESQL
+
+#include "database-postgresql.h"
+
+#include "log.h"
+#include "exceptions.h"
+#include "settings.h"
+
+/*
+ * IMPORTANT NOTICE
+ * Every conversion from non string to const char* values
+ * should be done by using an intermediate std::string
+ * buffer.
+ * If you don't use this intermediate, the conversion
+ * to std::string followed by conversion to const char*
+ * could have bad issues, refering to bad memory areas.
+ */
+
+Database_PostgreSQL::Database_PostgreSQL(const Settings &conf):
+	m_connect_string(""), m_conn(NULL), m_pgversion(0)
+{
+	if (!conf.getNoEx("pgsql_connection", m_connect_string)) {
+		throw SettingNotFoundException("Set pgsql_connection string in world.mt to "
+			"use the postgresql backend\n"
+			"Notes:\n"
+			"pgsql_connection has the following form: \n"
+			"\tpgsql_connection = host=127.0.0.1 port=5432 user=mt_user password=mt_password dbname=minetest_world\n"
+			"mt_user should have CREATE TABLE, INSERT, SELECT, UPDATE and DELETE rights on the database.\n"
+			"Don't create mt_user as a SUPERUSER!");
+	}
+
+	connectToDatabase();
+}
+
+Database_PostgreSQL::~Database_PostgreSQL()
+{
+	PQfinish(m_conn);
+}
+
+void Database_PostgreSQL::connectToDatabase()
+{
+	m_conn = PQconnectdb(m_connect_string.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+
+	m_pgversion = PQserverVersion(m_conn);
+
+	if (m_pgversion < 90500) {
+		throw DatabaseException("PostgreSQL database error: "
+			"Server version 9.5 or greater required.");
+	}
+
+	infostream << "PostgreSQL Database: Version " << m_pgversion << " Connection made." << std::endl;
+}
+
+void Database_PostgreSQL::verifyDatabase()
+{
+	if (PQstatus(m_conn) == CONNECTION_OK)
+		return;
+
+	PQreset(m_conn);
+	ping();
+}
+
+void Database_PostgreSQL::ping()
+{
+	if (PQping(m_connect_string.c_str()) != PQPING_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+}
+
+bool Database_PostgreSQL::initialized() const
+{
+	return (PQstatus(m_conn) == CONNECTION_OK);
+}
+
+void Database_PostgreSQL::initStatements()
+{
+	prepareStatement("read_block",
+			"SELECT block_data FROM blocks "
+					"WHERE posX = $1::int AND posY = $2::int AND posZ = $3::int");
```

Comment:
It's postgresql not sqlite or leveldb. There is a primary key with a good index which permit to retrieve data very fast.
On my server i have a full generated database (every mapblock is generated)

```
epixel=# select count(*) from blocks;
  count   
----------
 10585104
```

The server responds instanly for each mapblock called

```
epixel=# \timing on
epixel=# select * from blocks where posx>1800 and posy=-1869 limit 1;
 posx | posy  | posz  |                                                                        block_data                                                                        | b_version | eb_version | b_flags | b_content_width | b_params_width | b_nodes | b_nodes_metas |           b_time           
------+-------+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+------------+---------+-----------------+----------------+---------+---------------+----------------------------
 1867 | -1869 | -1873 | \x19090202789cedc13101000000c2a0f54f6d0c1fa000000000000000000000000000000080b70140000001789c63000000010001000000ffffffff0000010000000669676e6f72650a0000 |        26 |          0 |       0 |               2 |              2 | \x      | \x            | 2015-11-28 15:42:36.190229
(1 row)

Time: 0.809 ms
```